Sources/NIOHTTP1/NIOTypedHTTPServerUpgradeHandler.swift

```diff
@@ -152,6 +152,17 @@ public final class NIOTypedHTTPServerUpgradeHandler<UpgradeResult: Sendable>: Ch
         }
     }
 
+    public func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+        switch event {
+        case let evt as ChannelEvent where evt == ChannelEvent.inputClosed:
+            // The remote peer half-closed the channel during the upgrade so we should close
+            context.close(promise: nil)
```

Comment:
Thanks for working on this!

Can we please go through the state machine here. In particular, if we are in the state unbuffering we should not close here but rather just hold the event until the unbuffering is done.

Can we also write a test for this and look if we need to do the same for the client upgrade handler?

Reply:
We should also consider only closing if we haven't received a full upgrade request, and otherwise we should buffer this.

Reply:
It should be possible to emulate half closure with the embedded channel by just sending the event to the channel.